funsor/tensor.py

```diff
@@ -0,0 +1,973 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import functools
+import itertools
+import warnings
+from collections import OrderedDict
+from contextlib import contextmanager
+from functools import reduce
+
+import numpy as np
+import opt_einsum
+import torch
+from multipledispatch import dispatch
+from multipledispatch.variadic import Variadic
+
+import funsor.ops as ops
+from funsor.delta import Delta
+from funsor.domains import Domain, bint, find_domain, reals
+from funsor.ops import GetitemOp, MatmulOp, Op, ReshapeOp
+from funsor.terms import (
+    Binary,
+    Funsor,
+    FunsorMeta,
+    Lambda,
+    Number,
+    Slice,
+    Unary,
+    Variable,
+    eager,
+    substitute,
+    to_data,
+    to_funsor
+)
+from funsor.util import getargspec, quote
+
+
+numeric_array = (torch.Tensor, np.ndarray, np.generic)
```

Comment:
NumPy scalars belong to the class [np.generic](https://docs.scipy.org/doc/numpy/reference/generated/numpy.generic.html) rather than `np.ndarray`, so we need to include them here.